test/lib/jdk/test/lib/Platform.java

```diff
@@ -125,6 +125,23 @@ public static boolean isBusybox(String tool) {
         }
     }
 
+    public static boolean isCoreutilsSingleExecutable(String tool) {
+        try {
+            ProcessBuilder pb = new ProcessBuilder("coreutils", "--help");
+            pb.redirectErrorStream(true);
+            Process p = pb.start();
+            BufferedReader b = new BufferedReader(new InputStreamReader(p.getInputStream()));
+            String line;
+            while ((line = b.readLine()) != null) {
+                if (line.contains(tool)) {
+                    return true;
+                }
+            }
+        } catch(Exception e) {
+        }
+        return false;
+    }
+
     public static boolean isOSX() {
```

Comment:
This seems like a real hack. Is there really no better way to identity how sleep is implemented.
I'd probably favor an alternative that accepts multiple values for the expected value.

Reply:
Just ignore the binary name and only check for argv1 is imho enough.

Reply:
I couldn't think if a better way to detect what the host system had and adding this seemed inline with the busybox check. 

The choice of running `sleep` seems like it was arbitrary and the test should use something that is under JDK control rather than a platform specific tool that can change from distro to distro. Is there any reason we don't use the `java`, something else in the image, or even something built as part of the test-image.

Reply:
To be fair, this looks similar to what `isMusl()` does: searching for strings in outputs. But I do wonder if `--help` would include "sleep" for some innocuous description somewhere, and this code would accidentally match. @lutkerd, what's the output for `coreutils --help` for a single executable binary?

Reply:
> what's the output for `coreutils --help` for a single executable binary?

```
# coreutils --help
Usage: coreutils --coreutils-prog=PROGRAM_NAME [PARAMETERS]... 
Execute the PROGRAM_NAME built-in program with the given PARAMETERS.

      --help     display this help and exit
      --version  output version information and exit

Built-in programs:
 [ arch b2sum base32 base64 basename basenc cat chcon chgrp chmod chown chroot cksum comm cp csplit cut date dd df dir dircolors dirname du echo env expand expr factor false fmt fold ginstall groups head hostid id join link ln logname ls md5sum mkdir mkfifo mknod mktemp mv nice nl nohup nproc numfmt od paste pathchk pinky pr printenv printf ptx pwd readlink realpath rm rmdir runcon seq sha1sum sha224sum sha256sum sha384sum sha512sum shred shuf sleep sort split stat stdbuf stty sum sync tac tail tee test timeout touch tr true truncate tsort tty uname unexpand uniq unlink users vdir wc who whoami yes

Use: 'coreutils --coreutils-prog=PROGRAM_NAME --help' for individual program help.

GNU coreutils online help: <https://www.gnu.org/software/coreutils/>
Report any translation bugs to <https://translationproject.org/team/>
Full documentation <https://www.gnu.org/software/coreutils/coreutils>
or available locally via: info '(coreutils) Multi-call invocation'
```

This test is about parameter checking, so I am inclined to agree with @ecki and not check the process name or change the exe.

Reply:
At the time `sleep` seemed ubiquitous and was native so it ran quickly. (Much quicker than running java).
Running another program would be fine.  For example ,selecting an executable based on the OS type and giving the expected string in the output would be preferred.
The isBusyBox check is also a bad hack, but lighter weight since its not executing a program.
As the number of minor variations of platforms appear a more general approach would be useful.

I'd code it all in the test itself, doing a lookup based on the operating system name with the respective executable and expected string in the info arguments[0].
(With a fallback to sleep, so the table does not have to list every os).

Reply:
Any reason not to use [exesanity_SimpleNativeLauncher](https://github.com/openjdk/jdk/blob/628cd8a489fd54db18204c3bbaf4339d7ab5e9d6/test/jdk/native_sanity/simplenativelauncher/exesanity_SimpleNativeLauncher.c) or [exeBasicSleep](https://github.com/openjdk/jdk/blob/628cd8a489fd54db18204c3bbaf4339d7ab5e9d6/test/jdk/java/lang/ProcessBuilder/exeBasicSleep.c) in all cases?

Reply:
I'd be ok with exeBasicSleep; (currently only used on Windows and if its been built by the make files).
Note the discovery mechanism used in Basic.java `initSleepPath()` to locate and identify the path where its built or to fallback to the OS sleep.

Reply:
I opted to just check for coreutils existing as a signal that it was single executable. With Fedora and AL2023 coreutils-single package this will work. Some other distro could choose to use the symlink version of this or exclude sleep from coreutils-single in the future, but we can deal with that if it every actually comes up.